core/apple/test/SourceNSInputStreamTest.kt

```diff
@@ -0,0 +1,354 @@
+            ) {
+                usleep(100_000U)
+            }
+        }
+        @Suppress("ExtractKtorModule")
+        val server = embeddedServer(CIO, port = port, host = "localhost") {
+            routing {
+                post {
+                    this.context.request.receiveChannel().readAvailable(1) {
+                        println(it.readText())
+                    }
+                }
+            }
+        }
+        server.start(false)
+        tryUpload(NSInputStream(data = "default".encodeToByteArray().toNSData()))
+        tryUpload(Buffer().apply { writeString("custom") }.asNSInputStream())
+        server.stop()
+    }
+
+    @Test
+    fun testRunLoopPolling() {
+        val runLoop = startRunLoop()
+
+        val opened = Mutex(true)
+        val read = atomic(0)
+        val end = Mutex(true)
+
+        val buffer = Buffer()
+        val input = buffer.asNSInputStream()
+        val sink = ByteArray(6)
+        input.delegate = object : NSObject(), NSStreamDelegateProtocol {
+            override fun stream(aStream: NSStream, handleEvent: NSStreamEvent) {
+                println("event = ${handleEvent.asString()} (${NSThread.currentThread.name})")
+                when (handleEvent) {
+                    NSStreamEventOpenCompleted -> opened.unlock()
+                    NSStreamEventHasBytesAvailable -> {
+                        sink.usePinned {
+                            assertEquals(3, input.read(it.addressOf(read.value).reinterpret(), 3U))
+                            read.value += 3
+                        }
+                    }
+                    NSStreamEventEndEncountered -> end.unlock()
+                }
+            }
+        }
+        input.scheduleInRunLoop(runLoop, NSDefaultRunLoopMode)
+        input.open()
+        runBlocking {
+            opened.lockWithTimeout()
+            end.lockWithTimeout()
+            buffer.writeString("abc")
+            end.lockWithTimeout()
+            assertEquals(3, read.value)
+            buffer.writeString("123")
+            end.lockWithTimeout()
+        }
+        assertEquals(6, read.value)
+        assertEquals("abc123", sink.decodeToString())
+        input.close()
+        CFRunLoopStop(runLoop.getCFRunLoop())
+    }
+
+    @Test
+    fun testRunLoopSwitch() {
+        val runLoop1 = startRunLoop("run-loop-1")
+        val runLoop2 = startRunLoop("run-loop-2")
+
+        fun consumeSwitching(input: NSInputStream, data: String) {
+            println("consumeSwitching(${input::class}, $data)")
+            val opened = Mutex(true)
+            val readLock = reentrantLock()
+            var read = 0
+            val completed = Mutex(true)
+
+            input.delegate = object : NSObject(), NSStreamDelegateProtocol {
+                val sink = ByteArray(data.length)
+                override fun stream(aStream: NSStream, handleEvent: NSStreamEvent) {
+                    readLock.withLock {
```

Comment:
These tests use a relatively complex locking scheme to ensure the ordering and correctness of different phases of the stream's processing. It would be nice to have a comment describing the expected phase transition and locking order.